services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,376 @@
+	sampleN    = 4   // we sample 1 in every sampleN depth points
+	maxThreads = 300 // we will run at most maxThreads goroutines
+)
+
+func init() {
+	resource.RegisterService(svision.API, model, resource.Registration[svision.Service, *ObstaclesDepthConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (svision.Service, error) {
+			attrs, err := resource.NativeConfig[*ObstaclesDepthConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerObstacleDepth(ctx, c.ResourceName(), attrs, actualR)
+		},
+	})
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *ObstaclesDepthConfig) Validate(path string) ([]string, error) {
+	deps := []string{}
+	if config.K < 1 || config.K > 50 {
+		return nil, errors.New("invalid K, pick an integer between 1 and 50 (10 recommended)")
+	}
+	if (config.Hmin >= config.Hmax) && config.Hmax != 0 {
+		return nil, errors.New("Hmin should be less than Hmax")
+	}
+	return deps, nil
+}
+
+func registerObstacleDepth(
+	ctx context.Context,
+	name resource.Name,
+	conf *ObstaclesDepthConfig,
+	r robot.Robot,
+) (svision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerObstacleDistanceDetector")
+	defer span.End()
+	if conf == nil {
+		return nil, errors.New("config for obstacle_depth cannot be nil")
+	}
+
+	// If you have no intrinsics,
+	if conf.intrinsics == nil {
+		r.Logger().Warn("obstacle depth started without camera's intrinsic parameters")
+		segmenter := buildObsDepthNoIntrinsics()
+		return svision.NewService(name, r, nil, nil, nil, segmenter)
+	}
+
+	// Use defaults if needed
+	if conf.Hmax == 0 {
+		conf.Hmax = defaultHmax
+	}
+	if conf.ThetaMax == 0 {
+		conf.ThetaMax = defaultThetamax
+	}
+
+	myObsDep := obsDepth{
+		hMin: conf.Hmin, hMax: conf.Hmax, sinTheta: math.Sin(conf.ThetaMax),
+		intrinsics: conf.intrinsics, returnPCDs: conf.ReturnPCDs, k: conf.K,
+	}
+	segmenter := myObsDep.buildObsDepthWithIntrinsics() // does the thing
+	return svision.NewService(name, r, nil, nil, nil, segmenter)
+}
+
+// buildObsDepthNoIntrinsics will return the shortest depth in the depth map as a Geometry point.
+func buildObsDepthNoIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		min, _ := dm.MinMax()
+
+		pt := spatialmath.NewPoint(r3.Vector{X: 0, Y: 0, Z: float64(min)}, "obstacle")
```

Comment:
I wasn't expecting us to include labels on these obstacles.

Reply:
Removed here and below. I'm not married to it